File: packages/libsql/ios/Plugin/Libsql.swift
import Foundation
import Libsql

@objc public class Libsql: NSObject {

    private var databases: [String: Database] = [:]
    private var connections: [String: Connection] = [:]
    private var transactions: [String: String] = [:]

    @objc public func connect(_ options: ConnectOptions, completion: @escaping (_ result: ConnectResult?, _ error: Error?) -> Void) throws {
        do {
            let connectionId = UUID().uuidString

            let database: Database

            if let url = options.url, let path = options.getPath(), let authToken = options.authToken {
                // Embedded replica mode
                database = try Database(
                    path: path,
                    url: url,
                    authToken: authToken
                )
            } else if let url = options.url, let authToken = options.authToken {
                // Remote-only mode
                database = try Database(
                    url: url,
                    authToken: authToken
                )
            } else if let path = options.getPath() {
                // Local-only mode
                database = try Database(path)
            } else {
                // In-memory database
                database = try Database(":memory:")
            }

            let connection = try database.connect()

            self.databases[connectionId] = database
            self.connections[connectionId] = connection

            let result = ConnectResult(connectionId: connectionId)
            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    @objc public func execute(_ options: ExecuteOptions, completion: @escaping (_ error: Error?) -> Void) throws {
        do {
            guard let connection = self.connections[options.connectionId] else {
                completion(LibsqlError.connectionNotFound)
                return
            }

            if let values = options.values, !values.isEmpty {
                guard let valueRepresentables = values as? [ValueRepresentable] else {
                    completion(LibsqlError.invalidValues)
                    return
                }
                try connection.execute(options.statement, valueRepresentables)
            } else {
                try connection.execute(options.statement)
            }

            completion(nil)
        } catch {
            completion(error)
        }
    }

    @objc public func executeBatch(_ options: ExecuteBatchOptions, completion: @escaping (_ error: Error?) -> Void) throws {
        do {
            guard let connection = self.connections[options.connectionId] else {
                completion(LibsqlError.connectionNotFound)
                return
            }

            for (index, statement) in options.statement.enumerated() {
                let values = options.values?[safe: index]

                if let values = values, !values.isEmpty {
                    guard let valueRepresentables = values as? [ValueRepresentable] else {
                        completion(LibsqlError.invalidValues)
                        return
                    }
                    try connection.execute(statement, valueRepresentables)
                } else {
                    try connection.execute(statement)
                }
            }

            completion(nil)
        } catch {
            completion(error)
        }
    }

    @objc public func query(_ options: QueryOptions, completion: @escaping (_ result: QueryResult?, _ error: Error?) -> Void) throws {
        do {
            guard let connection = self.connections[options.connectionId] else {
                completion(nil, LibsqlError.connectionNotFound)
                return
            }

            let rows: Rows
            let tableName = extractTableName(from: options.statement)

            guard let tableName = tableName else {
                throw LibsqlError.tableNameNotRecognized
            }

            let columns = try connection.query("PRAGMA table_info(\(tableName))")
            var columnNames: [String] = []

            while let column = columns.next() {
                let nameValue = try column.get(1)
                if case .text(let columnName) = nameValue {
                    columnNames.append(columnName)
                }
            }

            if let values = options.values, !values.isEmpty {
                guard let valueRepresentables = values as? [ValueRepresentable] else {
                    completion(nil, LibsqlError.invalidValues)
                    return
                }
                rows = try connection.query(options.statement, valueRepresentables)
            } else {
                rows = try connection.query(options.statement)
            }

            let result = QueryResult(rows: rows, columns: columnNames)
            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    @objc public func beginTransaction(_ options: BeginTransactionOptions, completion: @escaping (_ result: BeginTransactionResult?, _ error: Error?) -> Void) throws {
        do {
            guard let connection = self.connections[options.connectionId] else {
                completion(nil, LibsqlError.connectionNotFound)
                return
            }

            _ = try connection.execute("BEGIN TRANSACTION")

            let transactionId = UUID().uuidString
            self.transactions[transactionId] = options.connectionId

            let result = BeginTransactionResult(transactionId: transactionId)
            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    @objc public func commitTransaction(_ options: CommitTransactionOptions, completion: @escaping (_ error: Error?) -> Void) throws {
        do {
            guard let connection = self.connections[options.connectionId] else {
                completion(LibsqlError.connectionNotFound)
                return
            }

            guard self.transactions[options.transactionId] != nil else {
                completion(LibsqlError.transactionNotFound)
                return
            }

            _ = try connection.execute("COMMIT")
            self.transactions.removeValue(forKey: options.transactionId)
            completion(nil)
        } catch {
            completion(error)
        }
    }

    @objc public func rollbackTransaction(_ options: RollbackTransactionOptions, completion: @escaping (_ error: Error?) -> Void) throws {
        do {
            guard let connection = self.connections[options.connectionId] else {
                completion(LibsqlError.connectionNotFound)
                return
            }

            guard self.transactions[options.transactionId] != nil else {
                completion(LibsqlError.transactionNotFound)
                return
            }

            _ = try connection.execute("ROLLBACK")
            self.transactions.removeValue(forKey: options.transactionId)
            completion(nil)
        } catch {
            completion(error)
        }
    }

    @objc public func sync(_ options: SyncOptions, completion: @escaping (_ error: Error?) -> Void) throws {
        do {
            guard let database = self.databases[options.connectionId] else {
                completion(LibsqlError.databaseNotFound)
                return
            }

            try database.sync()
            completion(nil)
        } catch {
            completion(error)
        }
    }

    private func extractTableName(from query: String) -> String? {
        let pattern = #"(?i)SELECT\s+\*\s+FROM\s+([a-zA-Z0-9_]+)"#

        if let regex = try? NSRegularExpression(pattern: pattern) {
            let range = NSRange(query.startIndex..., in: query)
            if let match = regex.firstMatch(in: query, range: range),
               let tableRange = Range(match.range(at: 1), in: query) {
                return String(query[tableRange])
            }
        }
        return nil
    }
}

extension Array {
    subscript(safe index: Index) -> Element? {
        return indices.contains(index) ? self[index] : nil
    }
}

File: packages/libsql/android/src/main/java/io/capawesome/capacitorjs/plugins/libsql/Libsql.kt
package io.capawesome.capacitorjs.plugins.libsql

import io.capawesome.capacitorjs.plugins.libsql.classes.options.*
import io.capawesome.capacitorjs.plugins.libsql.classes.results.*
import io.capawesome.capacitorjs.plugins.libsql.interfaces.*
import tech.turso.libsql.Database
import tech.turso.libsql.EmbeddedReplicaDatabase
import java.io.File
import tech.turso.libsql.Connection as LibsqlConnection
import java.util.concurrent.ConcurrentHashMap
import java.util.UUID

class Libsql(private val plugin: LibsqlPlugin) {
    
    private val databases = ConcurrentHashMap<String, Database>()
    private val connections = ConcurrentHashMap<String, LibsqlConnection>()
    private val transactions = ConcurrentHashMap<String, String>()

    @Throws(Exception::class)
    fun connect(options: ConnectOptions, callback: NonEmptyResultCallback<ConnectResult>) {
        try {
            val connectionId = UUID.randomUUID().toString()
            
            val database = when {
                options.url != null && options.path != null -> {
                
                    tech.turso.libsql.Libsql.open(
                        url = options.url!!,
                        authToken = options.authToken!!,
                        path = resolvePath(options.path!!)
                    )
                }
                options.url != null -> {
                
                    tech.turso.libsql.Libsql.open(
                        url = options.url!!,
                        authToken = options.authToken!!
                    )
                }
                options.path != null -> {
                
                    tech.turso.libsql.Libsql.open(path = resolvePath(options.path!!))
                }
                else -> {
                    
                    tech.turso.libsql.Libsql.open(path = ":memory:")
                }
            }

            val connection = database.connect()
            
            databases[connectionId] = database
            connections[connectionId] = connection

            val result = ConnectResult(connectionId)
            callback.success(result)
        } catch (exception: Exception) {
            callback.error(exception)
        }
    }

    @Throws(Exception::class)
    fun execute(options: ExecuteOptions, callback: EmptyCallback) {
        try {
            val connection = connections[options.connectionId]
                ?: throw Exception("Connection not found: ${options.connectionId}")

            val values = options.values?.toTypedArray()
            
            if (values != null && values.isNotEmpty()) {
                connection.execute(options.statement, *values)
            } else {
                connection.execute(options.statement)
            }

            callback.success()
        } catch (exception: Exception) {
            callback.error(exception)
        }
    }

    @Throws(Exception::class)
    fun executeBatch(options: ExecuteBatchOptions, callback: EmptyCallback) {
        try {
            val connection = connections[options.connectionId]
                ?: throw Exception("Connection not found: ${options.connectionId}")

            for (i in options.statement.indices) {
                val statement = options.statement[i]
                val values = options.values?.getOrNull(i)?.toTypedArray()
                
                if (!values.isNullOrEmpty()) {
                    connection.execute(statement, *values)
                } else {
                    connection.execute(statement)
                }
            }

            callback.success()
        } catch (exception: Exception) {
            callback.error(exception)
        }
    }

    @Throws(Exception::class)
    fun query(options: QueryOptions, callback: NonEmptyResultCallback<QueryResult>) {
        try {
            val connection = connections[options.connectionId]
                ?: throw Exception("Connection not found: ${options.connectionId}")

            val values = options.values?.toTypedArray()
            
            val rows = if (values != null && values.isNotEmpty()) {
                connection.query(options.statement, *values)
            } else {
                connection.query(options.statement)
            }

            val result = QueryResult(rows)
            callback.success(result)
        } catch (exception: Exception) {
            callback.error(exception)
        }
    }

    @Throws(Exception::class)
    fun beginTransaction(options: BeginTransactionOptions, callback: NonEmptyResultCallback<BeginTransactionResult>) {
        try {
            val connection = connections[options.connectionId]
                ?: throw Exception("Connection not found: ${options.connectionId}")

            connection.execute("BEGIN TRANSACTION")
            
            val transactionId = UUID.randomUUID().toString()
            transactions[transactionId] = options.connectionId

            val result = BeginTransactionResult(transactionId)
            callback.success(result)
        } catch (exception: Exception) {
            callback.error(exception)
        }
    }

    @Throws(Exception::class)
    fun commitTransaction(options: CommitTransactionOptions, callback: EmptyCallback) {
        try {
            val connection = connections[options.connectionId]
                ?: throw Exception("Connection not found: ${options.connectionId}")

            if (!transactions.containsKey(options.transactionId)) {
                throw Exception("Transaction not found: ${options.transactionId}")
            }

            connection.execute("COMMIT")
            transactions.remove(options.transactionId)

            callback.success()
        } catch (exception: Exception) {
            callback.error(exception)
        }
    }

    @Throws(Exception::class)
    fun rollbackTransaction(options: RollbackTransactionOptions, callback: EmptyCallback) {
        try {
            val connection = connections[options.connectionId]
                ?: throw Exception("Connection not found: ${options.connectionId}")

            if (!transactions.containsKey(options.transactionId)) {
                throw Exception("Transaction not found: ${options.transactionId}")
            }

            connection.execute("ROLLBACK")
            transactions.remove(options.transactionId)

            callback.success()
        } catch (exception: Exception) {
            callback.error(exception)
        }
    }

    
    @Throws(Exception::class)
    fun sync(options: SyncOptions, callback: EmptyCallback) {
        try {
            val database = databases[options.connectionId]
                ?: throw Exception("Database not found: ${options.connectionId}")

    
            val embeddedDb = database as? EmbeddedReplicaDatabase
                ?: throw Exception("Sync is only available for embedded replica databases")

            embeddedDb.sync()
            callback.success()
        } catch (exception: Exception) {
            callback.error(exception)
        }
    }
    

    private fun resolvePath(path: String): String {
        var file = File(path)
        if (!file.isAbsolute) {
            file = File(plugin.context.getDatabasePath(path).absolutePath)
        }
        return file.absolutePath
    }
}